### PR TITLE
nixos/keyd: add `extraConfig` option

### LINKED ISSUE #258083: keyd configuration is order-sensitive but the order is not being preserved
Composite layers in keyd MUST always be defined after regular layers (see
https://github.com/rvaiya/keyd/blob/6dc2d5c4ea76802fd192b143bdd53b1787fd6deb/docs/keyd.scdoc#L235 )

Currently it seems to be impossible to keep the settings order

@woojiq

### PR DESCRIPTION
**Pull request by woojiq**
## Description of changes

Add the `keyboards.<name>.extraConfig` option, which allows you to create more complex configurations. Especially useful when you need to use composite layers. Ideally, the order of set attributes needs be preserved, which is currently impossible with nix.

Resolves https://github.com/NixOS/nixpkgs/issues/258083

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by woojiq** [comment]
cc @peterhoeg can u review this pr?

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -18761,6 +18761,11 @@
     githubId = 1595132;
     name = "Kranium Gikos Mendoza";
   };
+  woojiq = {
+    github = "woojiq";
+    githubId = 122799969;
+    name = "Woojiq";
+  };
```
Comment (cafkafk) [inline maintainers/maintainer-list.nix:18768]: This should be added in it's own commit, as the first commit in this PR, with the summary `maintainers: add <handle>`

**Review by cafkafk — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (woojiq) [reply maintainers/maintainer-list.nix:18768]: Done. Btw, thanks for looking at this PR.

**Comment by cafkafk** [comment]
Result of `nixpkgs-review pr 258189` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

**Comment by woojiq** [comment]
I'm sorry, what does it mean, is it good or bad? The test `nix-build nixos/tests/keyd.nix` is passing if run manually.

File: pkgs/tools/inputmethods/keyd/default.nix [hunk 2]
```diff
@@ -71,7 +71,7 @@ stdenv.mkDerivation {
   meta = with lib; {
     description = "A key remapping daemon for linux.";
     license = licenses.mit;
-    maintainers = with maintainers; [ peterhoeg ];
+    maintainers = with maintainers; [ peterhoeg woojiq ];
```
Comment (jian-lin) [inline pkgs/tools/inputmethods/keyd/default.nix:74]: I think it is better to put this into a separate commit.

File: nixos/modules/services/hardware/keyd.nix [hunk 3]
```diff
@@ -37,6 +37,20 @@ let
           See <https://github.com/rvaiya/keyd> how to configure.
         '';
       };
+
+      extraConfig = mkOption {
+        type = types.str;
```
Comment (jian-lin) [inline nixos/modules/services/hardware/keyd.nix:42]: Is `types.lines` better because many definitions can be merged?

File: nixos/modules/services/hardware/keyd.nix [hunk 4]
```diff
@@ -82,21 +96,23 @@ in
...
+          nameValuePair "keyd/${name}.conf" {
+            text =
+              let
+                settings = settingsFormat.generate "keyd-${name}.conf" options.settings;
+              in
+              ''
                 [ids]
                 ${concatStringsSep "\n" options.ids}
+
+                ${builtins.readFile settings}
```
Comment (jian-lin) [inline nixos/modules/services/hardware/keyd.nix:111]: This is an [IFD](https://nixos.wiki/wiki/Import_From_Derivation).

I am not sure if it is fine to use IFD in NixOS modules.  I find an [example](https://github.com/NixOS/nixpkgs/pull/189587) which removes its usage.

**Comment by woojiq** [comment]
God, I messed up git:) I'll create another pr. I'm sorry for pinging.
